### ISSUE TITLE
[feature request] sparse x dense bmm

### ISSUE THREAD

**Issue by juniorrojas** [pytorch/pytorch#5672]
I'm working on a model that requires computing what's currently implemented in `torch.bmm`, with the difference that the first argument is a batch of sparse matrices rather than dense matrices. Since `torch.bmm` has nice autodiff and gpu support, it's the closest operation I found in pytorch that fits my current needs and I'm converting my sparse tensors to dense tensors to use it, but it's not optimal and it'd be very useful to have a sparse x dense bmm.

cc @ezyang @gchanan @zou3519 @jerryzh168 @vincentqb

**Comment by zou3519** [comment]
@atulkum Let me know if you have any specific questions. The biggest challenge here I think is to define what it means to have a "batch" of sparse tensors. Our sparse tensors have sparse dimensions and dense dimensions; but the sparse dimensions always come before the dense dimensions. Ideally we'd like the batch dimension (or dimensions) to be dense and at the beginning of the tensor, like elsewhere in PyTorch, but I'm not sure if our current sparse tensor api lets us do this.

**Comment by juniorrojas** [comment]
This is an example that computes a 2x2x2 sparse (`a`) - dense (`b`) bmm using a python array to store the sparse matrices. The expected output is a dense tensor. As @zou3519 said, the ideal implementation would encode the batch of sparse matrices using a pytorch tensor where only the first dimension is dense, but I'm not sure either if pytorch currently supports that.

```py
import torch

a1 = torch.sparse.FloatTensor(
    torch.LongTensor([
        [0, 1],
        [0, 1]
    ]),
    torch.FloatTensor([
        3, 4
    ]),
    torch.Size([2, 2])
)
a2 = torch.sparse.FloatTensor(
    torch.LongTensor([
        [0, 1],
        [1, 0]
    ]),
    torch.FloatTensor([
        2, 1
    ]),
    torch.Size([2, 2])
)
a = [a1, a2]

b = torch.FloatTensor([
    [
        [2, 1],
        [1, 4]
    ],
    [
        [3, -4],
        [1, -9]
    ]
])

ab = torch.stack([ai.mm(bi) for ai, bi in zip(a, b)])
print(ab)
```

**Comment by SsnL** [comment]
Actually, this does not seem that hard with cusparse's `cusparseSpMM`

**Comment by knsong** [comment]
Is there update for this issue? I also need the same feature.

**Comment by zou3519** [comment]
I'm upgrading this to high priority because of the demand for this feature.

**Comment by knsong** [comment]
> I'm upgrading this to high priority because of the demand for this feature.

Looking forward to it!

**Comment by elliottzheng** [comment]
The sparse version of bmm would be helpful when implementing GNN or other network based on sparse matrix.

**Comment by kurtamohler** [comment]
I can start working on this.

**Comment by kurtamohler** [comment]
Just wanted to share my thoughts before I start implementing this, to make sure I'm doing the right thing.

Regarding how to define a batch of sparse matrices, we could accept a 3-D sparse tensor, just like how we're accepting a 3-D dense tensor. If we have a set of N 2-D tensors of identical shapes, the caller could torch.stack() them all into one sparse tensor.

If we go this route, then in @juniorrojas 's example, instead of doing `a = [a1, a2]`, we could do `a = torch.stack((a1, a2))`. Then we would call `ab = torch.bmm(a, b)`, after I implement the proper overload for bmm.

This seems to be the most straightforward way to define a batch of sparse matrices, unless pytorch currently allows the first sparse tensor dimension to be dense. But @zou3519 mentioned above that the api might not support that.

But in either case, whether the first dimension is sparse or dense, it looks like we need a version of bmm() where the first arg is a 3-D sparse matrix and the second is a 3-D dense matrix. So I'll start working on that.

**Comment by kurtamohler** [comment]
Since the output of mm(sparse, dense) is a dense matrix, then I assume the output of bmm(sparse, dense) should also be dense.

**Comment by kurtamohler** [comment]
I made a CPU implementation of sparse times dense bmm. I focused on functionality before performance, and I figured the quickest way to implement this was to have it loop over dimension 0 of the input tensors, call `at::_sparse_mm()` on each pair of matrices, store the results in a vector, and then combine them all together with `at::stack_out()`.

It would be faster to output all of the matrix multiply results directly into the result matrix, avoiding the `at::stack_out()` call. Furthermore, I'm guessing that `SparseTensor::operator[]()` is a fairly expensive operation--it would probably be better to iterate through each of the index-value pairs.

But I'll do some performance comparisons between my current implementation and juniorrojas's workaround to see where it really stands.

**Comment by kurtamohler** [comment]
I wrote the following performance measurement script to compare my implementation with @juniorrojas 's workaround. In both cases, I am measuring the time it takes to calculate bmm given a list of 2D sparse matrices (since that is what juniorrojas starts with) and one 3D dense tensor.
```
import torch
import random
import time

# modify these values to perform different experiments:
num_matrices = 10000
dim_size = 100  # for simplicity, we use square matrices
sparsity = 0.99
timed_iters = 1

# Create a sparse matrix of specified sparsity and size, with
# randomized nonzero element positions and values
def get_rand_sparse_matrix(dim0, dim1, sparsity):
    size = [dim0, dim1]
    matrix = torch.zeros(size, dtype=torch.float)
    num_elements = dim0 * dim1
    num_nnz = max(int(num_elements * (1-sparsity)), 1)

    nnz_inds_1D = random.sample(range(num_elements), num_nnz)
    for nnz_ind_1D in nnz_inds_1D:
        nnz_ind_2D = (
            int(nnz_ind_1D % dim0),
            int(nnz_ind_1D / dim0)
        )
        rand_num = random.random()
        while rand_num == 0:
            rand_num = random.random()

        matrix[nnz_ind_2D] = rand_num

    return matrix.to_sparse()

def calc_bmm_builtin(a, b):
    a_stacked = torch.stack(a)

    ab = a_stacked.bmm(b)

    return ab

def calc_bmm_workaround(a, b):

    ab = torch.stack([ai.mm(bi) for ai, bi in zip(a, b)])
    return ab

def measure_bmm_time(num_matrices, a_dim0, a_dim1, b_dim1, sparsity, timed_iters):
    b_dim0 = a_dim1
    a = []

    for mat_ind in range(num_matrices):
        a.append(get_rand_sparse_matrix(a_dim0, a_dim1, sparsity))

    b = torch.rand([num_matrices, b_dim0, b_dim1], dtype=torch.float)

    # do one warmup loop
    for _ in range(2):
        start_workaround = time.time()
        for _ in range(timed_iters):
            ab_workaround = calc_bmm_workaround(a, b)
        time_workaround = (time.time() - start_workaround) / timed_iters

        start_builtin = time.time()
        for _ in range(timed_iters):
            ab_builtin = calc_bmm_builtin(a, b)
        time_builtin = (time.time() - start_builtin) / timed_iters

        # make sure the calculations match
        if not ab_workaround.eq(ab_builtin).all():
            print("outputs don't match!")
            exit(1)

    return time_workaround, time_builtin

time_workaround, time_builtin = measure_bmm_time(num_matrices, dim_size, dim_size, dim_size, sparsity, timed_iters)

print("workaround: %f s" % time_workaround)
print("builtin:    %f s" % time_builtin)
```

At the moment, my implementation is significantly slower than the workaround. Using inputs of size N x N x N, my implementation usually takes about twice the time of the workaround, for all size and sparsity combinations I tried.

Using inputs of size M x N x N, the speedup of the workaround versus my implementation seems to be roughly proportional to M / N. In other words, the more 2D matrix multiplies we do, the worse my implementation's performance gets. This makes sense because the more matrices we use, the more non-parallelized loops we're executing inside bmm() and the more times we're calling `SparseTensor::operator[]()`, which is probably pretty expensive. So I will focus on fixing these performance issues.

**Comment by ezyang** [comment]
> Regarding how to define a batch of sparse matrices, we could accept a 3-D sparse tensor, just like how we're accepting a 3-D dense tensor. If we have a set of N 2-D tensors of identical shapes, the caller could torch.stack() them all into one sparse tensor.

> If we go this route, then in juniorrojas 's example, instead of doing a = [a1, a2], we could do a = torch.stack((a1, a2)). Then we would call ab = torch.bmm(a, b), after I implement the proper overload for bmm.

I agree.

> This seems to be the most straightforward way to define a batch of sparse matrices, unless pytorch currently allows the first sparse tensor dimension to be dense. But zou3519 mentioned above that the api might not support that.

Richard is correct. Our format for COO sparse tensors does not support leading dense dimension.

> So I will focus on fixing these performance issues.

It might be a good idea to first do the CUDA implementation (where cusparse provides an efficient implementation) before we start investing in the CPU implementation.

**Comment by ezyang** [comment]
Looking at your posted commit, you're probably getting slaughtered by the indexing operation on sparse tensors. If you coalesce the sparse tensor first you can assume that the indices are in sorted order and you can project out views on the sparse tensor without having to move everything around.

**Comment by hongzimao** [comment]
Bump. Just to check in to see if this feature is available somewhere? We are working on a research project and are keen to work with any nightly built code for this feature (i.e., bmm(sparse, dense) = dense). @kurtamohler is something available from your local repo?

**Comment by kurtamohler** [comment]
@hongzimao , my most recent changes are here: https://github.com/kurtamohler/pytorch/tree/bmm-sparse-dense-5672

But shown in my comments above, my implementation is still slower than the workaround, so you're better off using the workaround for now.

**Comment by kurtamohler** [comment]
`cusparseSpMM()` is just a single sparse MxK times dense KxN multiply. It's not a batched implementation, so my CUDA implementation won't just be a wrapper for one cusparse function call. I think it will be just as (if not more) difficult for me to figure out how to implement the CUDA version of BMM as it will be to implement the CPU version. So I think I should focus on CPU first.

@ezyang, I've found out how to coalesce a sparse matrix, but I don't know what you mean by "project out views on the sparse tensor". Could you clarify and maybe point me to an example or an API that does this?

**Comment by kurtamohler** [comment]
I was able to improve the performance of the CPU implementation. I mimicked how the non-batched sparse x dense operation works, but I added in the third dimension. Performance is about 25% worse than the BMM workaround, but that's much better than it was before. I'll keep messing with things to find out what I can do to improve it further.

**Comment by ezyang** [comment]
OK, starting with CPU sounds reasonable. We might be able to loop over cusparseSpMM ourselves (in a CUDA kernel).

> "project out views on the sparse tensor"

Basically you can use `x._indices()` and `x._values()` to look at the index/value tensors on sparse

**Comment by kurtamohler** [comment]
@ezyang , I'm not sure I understand what coalescing does. Does it just sort everything in the sparse tensor by its index coordinates? I've noticed that my performance increases significantly if I don't coalesce, giving me better performance than the workaround, which tells me that coalescing is not necessary for the specific tensors I'm using. What situations would cause a sparse tensor to be uncoalesced? I'd like to see what my performance is if I give BMM the "most uncoalesced" tensor possible.

**Comment by kurtamohler** [comment]
I've realized that it actually doesn't matter functionally whether or not the input sparse matrix is coalesced. Since I'm applying an addition to the output tensor for each value in the sparse matrix, we'll get the same result whether a given element is coalesced into one index-value pair or if it's split between  two. For instance, if we have a coalesced tensor with an element at some index, whose value is V, we'll end up doing the following:
```
corresponding row of output tensor += (corresponding row of dense input) * V
```
But now if the sparse tensor is not coalesced, and the value V is split between two index-value pairs of value a and value V-a, we'll end up doing the following:
```
corresponding row of output tensor += (corresponding row of dense input) * a
corresponding row of output tensor += (corresponding row of dense input) * (V-a)
```
So the coalesced versus non-coalesced versions are mathematically equivalent, except that the non-coalesced version will perform more operations. However, since it can take a long time to coalesce a large tensor, it can be more desirable to avoid coalescing, depending on how many duplicate indices the tensor has. In fact, I've found that if I avoid coalescing and run my performance measurement script, my BMM implementation is faster than the workaround, for the specific tensors used in the script. So in my opinion, it could make sense to avoid coalescing in the BMM operation, allowing the user to make the decision depending on what gives better performance for their specific sparse tensor.

@ezyang, does that sound reasonable to you?

**Comment by ezyang** [comment]
Yes, the suggestion to coalesce was specifically under the assumption that you wanted to make repeated calls to a black box un-batched sparse-dense bmm. If you're going to straight up reimplement the sparse-dense matrix multiply (I'm not actually too sure how wise this is; it depends on how good/bad our non-batched mm is), then you may indeed be able to do things without coalescing.

**Comment by kurtamohler** [comment]
@ezyang, I see. I'll try to find an efficient way to call the unbatched sparse-dense bmm. That would be better to do it that way if I can make it fast enough.

**Comment by kurtamohler** [comment]
I was able to make a fairly performant implementation where I'm calling the original un-batched sparse x dense operation for each set of indices and values. In some cases it gives a little bit better performance than the workaround, but in most cases I ran, I'm seeing about a 2x slowdown.

I believe one of the bottlenecks is that I'm iterating through all of the sparse tensor's indices in order to determine the correct range of index-value pairs to send to each un-batched MM call. This task has a runtime of `O(M*N)`, where M=number of matrices, N=avg number of non-zero elements in each matrix. The workaround does not need to perform this task because the sparse matrices are already separated from each other as a list of tensors, which is why I believe this is one of the main performance differences between my implementation and the workaround.

I should be able to improve my method of determining the ranges by using a binary search, because I'm just looking at a sorted list that looks like `[0, 0, 0, 0, ..., 1, 1, 1, 1, ..., 2, 2, 2, 2, ..., ...]` to figure out where all the 0's are, where all the 1's are, etc. Using a binary search can make this task's runtime `O(M*log(M*N))`, which is usually going to be much better than `O(M*N)`, as shown by this plot of the function `M*N - M*log(M*N)`: https://www.wolframalpha.com/input/?i=m*n+-+m*log%28m*n%29+plot+for+0+%3C%3D+n+%3C%3D+1000%2C+0+%3C%3D+m+%3C%3D+1000

An easy improvement, once I have the binary search working, is to decrease the range over which I'm searching for each subsequent matrix. Since I know that I don't have to search within the ranges corresponding to the previous matrices, I can just exclude them from the search.

**Comment by ezyang** [comment]
Yep, binary search seems like the right thing to do here.

**Comment by kurtamohler** [comment]
I've added the binary search to find each sparse matrix's elements, and the performance did improve as expected. Using the performance measurement script below, I've compared the performance of my implementation to the workaround.

```
# Compares performance of torch.bmm(sparse, dense) with
# a workaround that was used in the absense of this function
# Since performance of sparse tensor operations is highly
# dependent on the number of nonzero elements, we have to
# run the measurement while varying both tensor sizes
# and percentages of nonzeros.

import torch
import random
import time
import math

torch.set_num_threads(1)

# Create a sparse matrix of specified sparsity and size, with
# randomized nonzero element positions and values
def get_rand_sparse_matrix(dim0, dim1, sparsity, device):
    size = [dim0, dim1]
    matrix = torch.zeros(size, dtype=torch.float, device=device)
    num_elements = dim0 * dim1
    num_nnz = max(int(num_elements * (1-sparsity)), 1)

    nnz_inds_1D = random.sample(range(num_elements), num_nnz)
    for nnz_ind_1D in nnz_inds_1D:
        nnz_ind_2D = (
            int(nnz_ind_1D % dim0),
            int(nnz_ind_1D / dim0)
        )
        rand_num = random.random()
        while rand_num == 0:
            rand_num = random.random()

        matrix[nnz_ind_2D] = rand_num

    return matrix.to_sparse()

def calc_bmm_builtin(a, b):
    ab = a.bmm(b)
    return ab

def calc_bmm_workaround(a, b):
    ab = torch.stack([ai.mm(bi) for ai, bi in zip(a, b)])
    return ab

# Measure the average time to calculate BMM on a randomized set of tensors
# using three methods: builtin BMM, builtin BMM given a pre-coalesced sparse tensor,
# and a workaround to calculate BMM without the builtin method
def measure_bmm_time(num_matrices, a_dim0, a_dim1, b_dim1, sparsity, timed_iters, device):
    b_dim0 = a_dim1
    a = []

    for mat_ind in range(num_matrices):
        a.append(get_rand_sparse_matrix(a_dim0, a_dim1, sparsity, device).coalesce())

    a_stacked = torch.stack(a)
    a_stacked_coalesced = torch.stack(a).coalesce()
    b = torch.rand([num_matrices, b_dim0, b_dim1], dtype=torch.float, device=device)

    # do one warmup loop
    for _ in range(2):
        start_workaround = time.time()
        for _ in range(timed_iters):
            ab_workaround = calc_bmm_workaround(a, b)
        time_workaround = (time.time() - start_workaround) / timed_iters

        start_builtin = time.time()
        for _ in range(timed_iters):
            ab_builtin = calc_bmm_builtin(a_stacked, b)
        time_builtin = (time.time() - start_builtin) / timed_iters

        start_builtin_coalesced = time.time()
        for _ in range(timed_iters):
            ab_builtin_coalesced = calc_bmm_builtin(a_stacked_coalesced, b)
        time_builtin_coalesced = (time.time() - start_builtin_coalesced) / timed_iters

        # make sure the calculations match
        if not ab_workaround.eq(ab_builtin).all():
            print("ab_builtin doesn't match ab_workaround!")
            exit(1)
        if not ab_workaround.eq(ab_builtin_coalesced).all():
            print("ab_builtin_coalesced doesn't match ab_workaround!")
            exit(1)

    return time_workaround, time_builtin, time_builtin_coalesced

device = 'cpu'

print("num_matrices squ_mat_elements sparsity workaround builtin builtin-precoalesced")

for num_matrices in [10, 100, 1000, 10_000]:
    for dim_size in [10, 100, 1000]:
        for sparsity in [0.999, 0.99, 0.9, 0]:
            nnz = int(num_matrices*dim_size*dim_size*(1.-sparsity))

            if nnz <= 100_000 and nnz > num_matrices:
                # The number of iterations to average over is inversely
                # proportional to the number of elements in the sparse tensor.
                # This is because a smaller sparse tensor will take less time,
                # so we'll want to run more iterations to get a more stable
                # measurement.
                timed_iters = int(math.ceil(5_000./nnz))
                time_workaround, time_builtin, time_builtin_coalesced = measure_bmm_time(
                    num_matrices,
                    dim_size, dim_size, dim_size,
                    sparsity,
                    timed_iters,
                    device
                )
                print("%d %d %f %f %f %f" % (num_matrices, dim_size**2, sparsity, time_workaround, time_builtin, time_builtin_coalesced))
```
This script runs the workaround and the builtin (my implementation) BMM on a single core with many different input sizes and sparsities. I'm running the builtin BMM with an un-coalesced sparse tensor so that the BMM operation has to run `coalesce()`, and then once again with a pre-coalesced sparse tensor so that the BMM does not have to run `coalesce()`. This way I can isolate the extra time it takes to coalesce.

I ran this script 5 times, calculated the speedups of builtin-vs-workaround and builtin_precoalesced-vs-workaround for each measurement, and then averaged the 5 sets of speedups. Here's what I get:

num_matrices | squ_mat_elements | sparsity | nnz | avg speedup | avg precoalesced speedup
-- | -- | -- | -- | -- | --
10 | 100 | 0.9 | 100 | 1.839725362 | 2.862684966
10 | 100 | 0 | 1000 | 1.175196548 | 2.671741036
10 | 10000 | 0.999 | 100 | 1.515473223 | 1.909875557
10 | 10000 | 0.99 | 1000 | 1.184215142 | 1.924977751
10 | 10000 | 0.9 | 10000 | 0.4613916705 | 1.521560519
10 | 10000 | 0 | 100000 | 0.1838851067 | 1.150508587
10 | 1000000 | 0.999 | 10000 | 1.172411347 | 1.412706636
10 | 1000000 | 0.99 | 100000 | 0.7613480796 | 1.090194927
100 | 100 | 0.9 | 1000 | 2.75806101 | 3.212183026
100 | 100 | 0 | 10000 | 1.452760009 | 2.838050521
100 | 10000 | 0.999 | 1000 | 1.835663576 | 1.975791373
100 | 10000 | 0.99 | 10000 | 1.345027544 | 1.885943046
100 | 10000 | 0.9 | 100000 | 0.4604152852 | 1.597152826
100 | 1000000 | 0.999 | 100000 | 1.197337685 | 1.228813969
1000 | 100 | 0.9 | 10000 | 2.956959462 | 3.274028691
1000 | 100 | 0 | 100000 | 1.487353866 | 2.934268352
1000 | 10000 | 0.999 | 10000 | 1.660517611 | 1.742181125
1000 | 10000 | 0.99 | 100000 | 1.178916995 | 1.567672849
10000 | 100 | 0.9 | 100000 | 3.010757125 | 3.341205249
10000 | 10000 | 0.999 | 100000 | 1.742236612 | 1.826596483

My implementation without pre-coalescing is faster than the workaround for all but 4 of the input sizes I used. But with pre-coalescing, I get a speedup in every case. So coalescing is taking a significant amount of time in some cases.

There are probably some more optimizations to be found, but I think this is acceptable performance for now. I'll start implementing the CUDA version, which I'm guessing will be very similar to the CPU version.

**Comment by ezyang** [comment]
> I'll start implementing the CUDA version, which I'm guessing will be very similar to the CPU version.

I would not bet on that :)

**Comment by kurtamohler** [comment]
I'm currently thinking about how to make my CUDA implementation faster. While my CUDA implementation is faster than the workaround, it's still slower than the CPU version in almost all cases.

I based my CUDA implementation on the CPU implementation so that I could get it into a state that at least gives the correct result. So it's copying the first row of the indices matrix over to a CPU tensor and binary searching through that to find where all the matrices begin and end. I think the copy is one of the main bottlenecks, and, furthermore, the searches aren't being parallelized. It would be much better if I could make a CUDA kernel that executes M (number of matrices) separate searches of the index array at the same time. One of the challenges here will probably be to implement a binary search kernel that doesn't have significant branch divergence. From what I've read about GPUs, it seems like ternary operators can often be used to generate conditional move instructions, rather than branches, so I will try that out.

**Comment by kurtamohler** [comment]
Does anyone here have experience working with Cusparse? I'm having trouble figuring out how to properly create a cusparseSpMatDescr_t structure, which is needed for the cusparseSpMM() function. I'm hoping that I can reuse the data from an at::SparseTensor, so I'm trying to use the cusparseCreateCoo() function with pointers into the at::SparseTensor.

I've written the call to cusparseSpMM(), but the result seems to always be a 0 matrix. So I'm wondering if my assumption that I can reuse the at::SparseTensor is wrong.

**Comment by ezyang** [comment]
Full disclosure: I haven't written any CUDA kernels in a long time, so I'm not the best person to ask for CUDA-side algorithmic advice.

**Comment by knsong** [comment]
@kurtamohler Did you have any progress on debugging your CUDA implements?

**Comment by kurtamohler** [comment]
@knsong , I think I'm close. I believe I've written proper cusparse calls to create matrices and to calculate the multiplication. But the result I get is incorrect, and I suspect that it has to do with the fact that ATen and cusparse have different representations of dense matrices in memory.

**Comment by kurtamohler** [comment]
I must be very close to getting this working, but I'm having trouble figuring out what's going wrong, so I think it's time for me to share what's going on. The above commit adds the `cusparseSpMM` call, but the result is incorrect most of the time.

I wrote a script to compare the output of this implementation with the output of the workaround using a few 2x2 matrices:

```
import torch
#torch.set_default_tensor_type('torch.cuda.FloatTensor')


a1 = torch.sparse.FloatTensor(
    torch.LongTensor([
        [0],
        [0],
    ]),
    torch.FloatTensor([
        1
    ]),
    torch.Size([2, 2])
).cuda()
a2 = torch.sparse.FloatTensor(
    torch.LongTensor([
        [0],
        [1],
    ]),
    torch.FloatTensor([
        1
    ]),
    torch.Size([2, 2])
).cuda()
a3 = torch.sparse.FloatTensor(
    torch.LongTensor([
        [1],
        [0],
    ]),
    torch.FloatTensor([
        1
    ]),
    torch.Size([2, 2])
).cuda()
a4 = torch.sparse.FloatTensor(
    torch.LongTensor([
        [1],
        [1],
    ]),
    torch.FloatTensor([
        1
    ]),
    torch.Size([2, 2])
).cuda()

a5 = torch.sparse.FloatTensor(
    torch.LongTensor([
        [0, 0],
        [0, 1],
    ]),
    torch.FloatTensor([
        1, 1
    ]),
    torch.Size([2, 2])
).cuda()

a6 = torch.sparse.FloatTensor(
    torch.LongTensor([
        [0, 1],
        [0, 0],
    ]),
    torch.FloatTensor([
        1, 1
    ]),
    torch.Size([2, 2])
).cuda()

a7 = torch.sparse.FloatTensor(
    torch.LongTensor([
        [0, 1],
        [1, 0],
    ]),
    torch.FloatTensor([
        2, 1
    ]),
    torch.Size([2, 2])
).cuda()

a8 = torch.sparse.FloatTensor(
    torch.LongTensor([
        [1, 0],
        [0, 1],
    ]),
    torch.FloatTensor([
        1, 2
    ]),
    torch.Size([2, 2])
).cuda()

a = [a1, a2, a3, a4, a5, a6, a7, a8]
a_stacked = torch.stack(a).cuda()



b = torch.FloatTensor(
    [[
        [1, 2],
        [3, 4]
    ]]*len(a)
).cuda()
print("a = " + str(a_stacked.to_dense()))
print("b = " + str(b))

ab = torch.stack([ai.mm(bi) for ai, bi in zip(a, b)])
print("a*b = " + str(ab))

ab_stacked = a_stacked.bmm(b)
print("a*b stacked = " + str(ab_stacked))
```

The script's output shows the correct answer if the sparse matrix only has one non-zero element, but incorrect answers when there are more than one non-zero element:

```
a = tensor([[[1., 0.],
         [0., 0.]],

        [[0., 1.],
         [0., 0.]],

        [[0., 0.],
         [1., 0.]],

        [[0., 0.],
         [0., 1.]],

        [[1., 1.],
         [0., 0.]],

        [[1., 0.],
         [1., 0.]],

        [[0., 2.],
         [1., 0.]],

        [[0., 2.],
         [1., 0.]]], device='cuda:0')
b = tensor([[[1., 2.],
         [3., 4.]],

        [[1., 2.],
         [3., 4.]],

        [[1., 2.],
         [3., 4.]],

        [[1., 2.],
         [3., 4.]],

        [[1., 2.],
         [3., 4.]],

        [[1., 2.],
         [3., 4.]],

        [[1., 2.],
         [3., 4.]],

        [[1., 2.],
         [3., 4.]]], device='cuda:0')
a*b = tensor([[[1., 2.],
         [0., 0.]],

        [[3., 4.],
         [0., 0.]],

        [[0., 0.],
         [1., 2.]],

        [[0., 0.],
         [3., 4.]],

        [[4., 6.],
         [0., 0.]],

        [[1., 2.],
         [1., 2.]],

        [[6., 8.],
         [1., 2.]],

        [[6., 8.],
         [1., 2.]]], device='cuda:0')
a*b stacked = tensor([[[ 1.,  2.],
         [ 0.,  0.]],

        [[ 3.,  4.],
         [ 0.,  0.]],

        [[ 0.,  0.],
         [ 1.,  2.]],

        [[ 0.,  0.],
         [ 3.,  4.]],

        [[ 2.,  4.],
         [ 0.,  0.]],

        [[ 2.,  4.],
         [ 0.,  0.]],

        [[ 7., 10.],
         [ 0.,  0.]],

        [[ 7., 10.],
         [ 0.,  0.]]], device='cuda:0')
```

**Comment by kurtamohler** [comment]
I think there's a bug in cusparseSpMM, although it's certainly possible that I'm still misunderstanding how to create cusparse matrices. I've made a standalone repo that shows the issue I'm seeing: https://github.com/kurtamohler/cusparseSpMM-COO-bug

I submitted a bug with NVIDIA: https://developer.nvidia.com/nvidia_bug/2846222

**Comment by ezyang** [comment]
Nice. Looking forward to see how NVIDIA replies.

**Comment by kurtamohler** [comment]
Liam from NVIDIA pointed out that I was using uint64_t indices in my sparse matrices, but the cusparseSpMM operation only supports 32-bit indices. Liam showed me that the docs do mention this, I just didn't read carefully enough :P Apparently, they might add 64-bit index support in the future.

In my standalone repo, I switched over to uint32_t indices and started getting correct results. So now I'll try fixing my BMM implementation to do the same thing. Hopefully the performance doesn't get killed by having to convert all the indices from 64-bit to 32-bit.

**Comment by kurtamohler** [comment]
Converting the sparse matrix indices to 32-bit fixed the issue! I'm now getting better performance with cuda than cpu for some of the cases that my performance script covers. The speedup isn't huge, and the cpu version still has an advantage for some input sizes, but at least the cuda version is now much faster than it was a few commits ago (used to be slower than cpu in every case).

I'm going to spend some time looking for any more potential optimizations. After that, I'll post a performance report.

**Comment by kurtamohler** [comment]
There are three optional COO sparse MM algorithms within cusparseSpMM, `CUSPARSE_COOMM_ALG1`, `_ALG2`, and `_ALG3`. For some reason, ALG3 fails--I think it has something to do with using too much workspace memory, but I'm not sure. I was using ALG2, which is deterministic. ALG1 is not deterministic, but I discovered that it's significantly faster than ALG2, by more than 4x in some cases.

I think what I should do is add a `bool deterministic` argument to my bmm functions to give the user the option to choose which of the two algorithms to use. What do you think @ezyang ?

**Comment by ezyang** [comment]
> I think what I should do is add a bool deterministic argument to my bmm functions to give the user the option to choose which of the two algorithms to use.

That is a reasonable stopgap, although actually exposing the flags to the user in a useful way is related to #15359. But a bool arg is better than nothing.

**Comment by kurtamohler** [comment]
Below are a couple plots of sparse-dense BMM's performance. The run times of CPU/CUDA, workaround/builtin, pre-coalesced/not-pre-coalesced are all included. The "input cases" that the x-axis refers to are the input size combinations found in the same order in the table below the charts.

The builtin methods are almost always faster than the workarounds, and CUDA is almost faster than CPU.

![BMM sparse-dense performance (not pre-coalesced)](https://user-images.githubusercontent.com/8378787/74689416-aef30200-51a0-11ea-8d47-f6123ed9c275.png)

![BMM sparse-dense performance (pre-coalesced)](https://user-images.githubusercontent.com/8378787/74689419-b4e8e300-51a0-11ea-98c5-4334b30e6b4c.png)

**Input cases:**

num_matrices | squ_mat_elements | output_size | sparsity | nnz
-- | -- | -- | -- | --
10 | 100 | 10000 | 0.9 | 100
10 | 100 | 100000 | 0 | 1000
10 | 10000 | 1000000 | 0.999 | 100
10 | 10000 | 10000000 | 0.99 | 1000
10 | 10000 | 100000000 | 0.9 | 10000
10 | 10000 | 1000000000 | 0 | 100000
10 | 1000000 | 10000000000 | 0.999 | 10000
10 | 1000000 | 100000000000 | 0.99 | 100000
100 | 100 | 100000 | 0.9 | 1000
100 | 100 | 1000000 | 0 | 10000
100 | 10000 | 10000000 | 0.999 | 1000
100 | 10000 | 100000000 | 0.99 | 10000
100 | 10000 | 1000000000 | 0.9 | 100000
100 | 1000000 | 100000000000 | 0.999 | 100000
1000 | 100 | 1000000 | 0.9 | 10000
1000 | 100 | 10000000 | 0 | 100000
1000 | 10000 | 100000000 | 0.999 | 10000
1000 | 10000 | 1000000000 | 0.99 | 100000
10000 | 100 | 10000000 | 0.9 | 100000
10000 | 10000 | 1000000000 | 0.999 | 100000

I think I'm basically done with everything I need to do at this point, so I'll create a PR now.

**Comment by ezyang** [comment]
@kurtamohler as a point of order, I suggest putting the benchmarks in the PR itself for better visibility ;)

**Comment by smr97** [comment]
Hi, I want to use this feature, could you please point me to a documentation/example preferrably that can show me how to to sparse batched multiplication?

**Comment by kurtamohler** [comment]
Hi @smr97, the documentation for `bmm` is here: https://pytorch.org/docs/stable/generated/torch.bmm.html?highlight=bmm#torch.bmm

In order to use the sparse-dense feature, the first argument (`input`) has to be a sparse tensor, and the second argument (`mat2`) has to be dense. For instance, the following will perform a batch of 10 multiplies of 4x5 sparse matrices by 5x6 sparse matrices.

```
torch.bmm(torch.randn(10, 4, 5).to_sparse(), torch.randn(10, 5, 6))
```

**Comment by smr97** [comment]
Thanks @kurtamohler, it was just tough to figure out that `bmm` is actually overloaded for SparseTensor types. This makes it clear. Thanks for the feature and the explanation!!

**Comment by atiorh** [comment]
Finding out about this awesome sparse `torch.bmm` support in torch is non-trivial. I believe that this should be exported/aliased under the `torch.sparse` namespace so that folks can learn about it in a central landing doc page.  Would you agree @kurtamohler ?

**Comment by kurtamohler** [comment]
@atiorh, yes I think that would be a great idea. I've opened an issue.

**Comment by ychfan** [comment]
@kurtamohler [Docs](https://docs.nvidia.com/cuda/cusparse/index.html#cusparse-generic-function-spmm) show that cusparseSpMM also supports batched computation. Do you think we can use it instead of loop?

**Comment by kurtamohler** [comment]
@ychfan, if you call `torch.bmm` with sparse-dense CUDA tensors, the operation is implemented with cusparseSpMM, as shown in my PR here: https://github.com/pytorch/pytorch/pull/33430/files#diff-98e035ee4319ebbd499f6be9da9af845R986

**Comment by ychfan** [comment]
@kurtamohler For my understanding, the current `bmm` is based a loop over batch axis and calling spmm for each matrix. Based on nv's doc, spmm can deal with batch in a single call.

**Comment by kurtamohler** [comment]
@ychfan , ah right, sorry I misunderstood. Yeah it looks like it should be possible. I don't remember seeing that in the cuda documentation when I worked on this, might be a relatively new feature

**Comment by kaiwenwu96** [comment]
@kurtamohler Thank you for your work on this function! If the memory is not an issue, then a simple method is converting all sparse tensors into dense ones then call the normal dense torch.bmm. Is there any comparison for the current torch.bmm implementation with this (possibly naive) one?

If I understand correctly, the current bmm implementation does a loop over the batch axis, which does not fully utilize GPUs. So I am just wondering when the speed-up by the sparsity dominates the sequential execution over the batch axis.

**Comment by kurtamohler** [comment]
@kaiwenwu96, converting sparse to dense and then performing a dense multiply might offer better performance for tensors smaller than a certain size and with a low enough sparsity. Although I'm not sure about the performance of the conversion itself, which could potentially nullify any performance gains, unless you can avoid the conversion by avoiding sparse tensors completely. Yes, the CUDA bmm implementation loops over the batch axis. I think you're right to ask if that significantly limits the usefulness of sparse-dense bmm for a specific application.

I imagine all of this is device-dependent. For one CPU or GPU, performance comparisons could be significantly different than those of a different CPU or GPU. I'm inclined to suggest that the most reliable comparison you could get is to measure it on your platform, adjusted for your specific application.

I hope that helps, I'm sorry I don't really have a definitive answer

**Comment by kurtamohler** [comment]
Related: I looked into cuSPARSE's batching feature for spmm. It only supports batches of matrices that all have the same number of nonzero elements. I'm not sure how useful this is--is it a common use case? If so, we can open a new issue to add support.

https://docs.nvidia.com/cuda/cusparse/index.html#cusparse-generic-spmat-set-coo-batches

**Comment by kaiwenwu96** [comment]
> @kaiwenwu96, converting sparse to dense and then performing a dense multiply might offer better performance for tensors smaller than a certain size and with a low enough sparsity. Although I'm not sure about the performance of the conversion itself, which could potentially nullify any performance gains, unless you can avoid the conversion by avoiding sparse tensors completely. Yes, the CUDA bmm implementation loops over the batch axis. I think you're right to ask if that significantly limits the usefulness of sparse-dense bmm for a specific application.
> 
> I imagine all of this is device-dependent. For one CPU or GPU, performance comparisons could be significantly different than those of a different CPU or GPU. I'm inclined to suggest that the most reliable comparison you could get is to measure it on your platform, adjusted for your specific application.
> 
> I hope that helps, I'm sorry I don't really have a definitive answer

@kurtamohler Thanks for the reply. You might be very much correct. It very likely depends on the specific use case and the sparsity level. In my case, I did the dense one. I constructed all tensors as dense matrices at the very beginning so the conversion is not an issue for me. The problem is, though, dense matrices consumes too much memory.  I recently discovered this functionality in pytorch so just want to try it. My matrices have sparsity level >= 99% so very likely it is going to save a lot of memory. The remaining concern is whether it hurt the running time. I will try this API to see whether it slows down or speed up in my case.
